crates/wasmtime/src/module/registry.rs

```diff
@@ -48,66 +43,91 @@ impl StoreFrameInfo {
         self.module(pc)?.lookup_trap_info(pc)
     }
 
-    fn module(&self, pc: usize) -> Option<&ModuleFrameInfo> {
-        let (end, info) = self.ranges.range(pc..).next()?;
+    /// Looks up a stack map from a program counter.
+    pub fn lookup_stack_map<'a>(&'a self, pc: usize) -> Option<&'a ir::StackMap> {
+        self.module(pc)?.lookup_stack_map(pc)
+    }
+
+    fn module(&self, pc: usize) -> Option<&RegisteredModule> {
+        let (end, info) = self.0.range(pc..).next()?;
         if pc < info.start || *end < pc {
             return None;
         }
 
         Some(info)
     }
 
-    /// Registers a new compiled module's frame information.
-    pub fn register(&mut self, module: &Arc<CompiledModule>) {
-        let (start, end) = module.code().range();
+    /// Registers a new module with the registry.
+    pub fn register(&mut self, module: &Module) {
+        let compiled_module = module.compiled_module();
+        let (start, end) = compiled_module.code().range();
 
-        // Ignore modules with no code or finished functions
-        if start == end || module.finished_functions().is_empty() {
+        // Ignore modules with no code, finished functions, or if the module is already registered
+        if start == end || compiled_module.finished_functions().is_empty() {
             return;
         }
 
         // The module code range is exclusive for end, so make it inclusive as it
         // may be a valid PC value
         let end = end - 1;
 
+        if self.0.get(&end).is_some() {
```

Comment:
Could this `debug_assert` that the start is the same?

Reply:
(and also have a comment as to why it's an early-return here for double-registration of a module.